docs/source/blueprints/BP002.rst

```diff
@@ -0,0 +1,220 @@
+ways to declare a requirement into a new representation. This new
+representation can serve as an input to one of several modules responsible for
+handling specific types of requirements, like, for example, files, images,
+packages, git repositories, source code or operating system parameters.
+
+Requirements representations
+============================
+
+Define how to represent a requirement is the first step to define the
+architecture of a resolver. This blueprint defines the following ways to
+represent a requirement:
+
+  1. JavaScript Object Notation (JSON) file;
+  2. Python executable that produces a JSON file;
+  3. metadata included in the test docstring.
+
+Requirements representation as JSON files
+-----------------------------------------
+
+JSON is a lightweight data-interchange format [2] supported by the Python
+standard library. Using it to represent requirements is flexible and
+straightforward.
+
+The standard proposed way to represent requirements with JSON is defining one
+requirement per entry. Each entry should start with the requirement type,
+followed by other keyword arguments related to that type. Example::
+
+        [
+            {"type": "url", "url": "https://cabort.com/cabort.c", "hash": "deadbeefdeadbeef"},
+            {"type": "vmimage", "distro": "fedora", "version": 31, "arch": "x86_64"},
+            {"type": "package", "package": "lvm"}
+        ]
+
+The requirement `type` should match the module responsible for that type of
+requirement.
+
+Requirements representation as Python executable
+------------------------------------------------
+
+Another way to create the requirements representation as JSON files is by
+writing a Python executable. This approach makes the requirements
+representation flexible, by allowing the use of Python variables and code that
+may change the parameters values for the requirements, depending on the
+environment the Python code runs.
+
+The following example shows a requirement that depends on the architecture
+the test is running::
+
+        #!/usr/bin/python3
+
+        import os
+        import json
+
+        requirements = [
+            {"type": "url", "url": "https://cabort.com/cabort.c", "hash": "deadbeefdeadbeef"},
+            {"type": "vmimage", "distro": "fedora", "version": 31, "arch": os.uname()[4]},
+            {"type": "package", "package": "lvm"}
+        ]
+
+        print(json.dumps(requirements))
+
+Requirements representation as metadata on test docstring
+---------------------------------------------------------
+
+Test writers may want to add the requirements of a test into de test code. The
+option proposed here allows the use of metadata on test docstrings to represent
+the requirements list.
+
+Below is an example of how to define requirements as metadata on docstrings::
+
+        def test_something(self):
+        '''
+            :avocado requirement: [{"type": "url", "url": "https://cabort.com/cabort.c", "hash": "deadbeefdeadbeef"}]
```

Comment:
This seems to corroborate with my previous point, in which the content is basically the same, except that there's the `:avocado: requirement` "envelop, which we can say is the substitute to the `test.data/requirements.json` "file envelop".

BTW, I think in this case that we can have multiple tags defined, we could default to list a single requirement per tag, and then, drop the list, that is:

```
:avocado: requirement={"type": "package", "package": "gcc"}
:avocado: requirement={"type": "package", "package": "lvm"}
```

Reply:
I struggled a lot trying to define the docstring format, so any idea here is appreciated and I liked yours.